### ISSUE TITLE
Guaranteeing logs are flushed - Best Practices

### ISSUE THREAD

**Issue by mhelbig-oz** [aws/aws-sdk-net#1675]
## The Question
Is there a proper way to guarantee that all log calls are written to Cloudwatch prior to a Lambda exiting? I see very rare cases where according to the logs, it appears that the Lambda just exits part way through. No exceptions are logged, and the logs I put in to flag success never make it to the file. This is flagged as an error by our internal processes, and every investigation I've done for these seem to indicate that everything did run successfully we just stopped logging partway through the invocation.

I am logging using Amazon.Lambda.Core.LambdaLogger.Log. If I use the logger on the ILambdaContext could it avoid this issue? Is there a flush I could call somewhere?

### Environment
* Package Version: Amazon.Lambda.Core 1.1
* OS Info: Amazon Linux
* Build Environment: Visual Studio
* Targeted .NET Platform:  

---

This is a :question: general question

**Comment by ashishdhingra** [comment]
Hi @mhelbig-oz,

Thanks for posting the guidance question. Please confirm if you are making any asynchronous call within the Lambda function handler which could cause the execution thread to pause while background long running process is complete. This may cause log entries to be out of order when function is invoked by multiple clients simultaneously.

Thanks,
Ashish

**Comment by mhelbig-oz** [comment]
Hi Ashish,

What I'm seeing is missing entries, not out of order entries. I do have async operations, but prior to shutdown all of the tasks are joined into the FunctionHandler thread. I then log out a final message indicating success prior to it exiting. This message and sometimes some before it are VERY rarely missing from the cloudwatch logs. Instead I just get the standard END and REPORT lines. There isn't an out of memory error or timeout which is another way that logging can just stop.

Thanks,
Matthew Helbig

**Comment by ashishdhingra** [comment]
Hi @mhelbig-oz,

Thanks for your reply. There is no specific pattern I could think of except that when the function handler resumes after async operation (may be using Task.Waitall()), log statements thereafter should be logged. It would be helpful if you could share your code snippet (with sensitive information removed) to see if something could be implemented differently.

Thanks,
Ashish

**Comment by mhelbig-oz** [comment]
Yes I saw that comment, I opened this bug because after reading it I still wasn't sure about a couple of things.

See the original post:
"I am logging using Amazon.Lambda.Core.LambdaLogger.Log. If I use the logger on the ILambdaContext could it avoid this issue?"

Also do I need to call LogLine instead of Log in order to guarantee delivery?

Thanks,
Matt

**Comment by ashishdhingra** [comment]
Hi @mhelbig-oz,

The only difference between Log() and LogLine() is that LogLine() method logs a message, followed by the current line terminator, to AWS CloudWatch Logs. (refer [ILambdaLogger](https://github.com/aws/aws-lambda-dotnet/blob/e2bb81a4b48a402570b0de828772067938ca5671/Libraries/src/Amazon.Lambda.Core/ILambdaLogger.cs#L18) interface).

Please also refer the question posted on [Stackoverflow](https://stackoverflow.com/questions/57586886/net-core-aws-lambda-function-not-flushing-logs) where it specifies the similar issue.

Kindly let me know if this provides some guidance and if we could close this issue.

Thanks,
Ashish

**Comment by mhelbig-oz** [comment]
No this is not helpful at all. As stated multiple times in this thread I am using the Amazon Libraries to log, specifically

**Amazon.Lambda.Core.LambdaLogger.Log**

This is your library, recommended from your documentation, not an external library as you've referenced multiple times. 

I am seeing missing logs extremely rarely. Yesterday out of 28,915 invocations I missed 2, so a reproduction rate of ~0.00007%. To me this smells like a rare threading issue within the logging library itself. I have many different lambdas but they all reproduce this behavior rarely, and they all have the same implementation of logging as one of the last things they do before shutting down. I have only seen missing logs at the very end of the invocation (too be fair that's what my automated tests are looking for). These logs are happening in the FunctionHandler callback immediately prior to the scope being exited.

**Comment by ashishdhingra** [comment]
Hi @mhelbig-oz,

I understand your concern but as stated in the documentation in [aws-logging-net](https://github.com/aws/aws-logging-dotnet#aws-lambda), since the logs are pushed to Cloudwatch via background thread, there is a slight chance that some of it might be missed. This could happen in a scenario where Lambda invocation is complete which pauses the background thread used for logs and that thread doesn't get a chance to resume if there is no Lambda invocation within a certain period of time. This could be the reason for the issue with the reproduction rate of 0.00007%. The response quoted in Stackoverflow by one of our senior developers confirms this limitation.

Thanks,
Ashish

**Comment by mhelbig-oz** [comment]
Hey Ahish,

The response by your senior developer is talking about using AWS.Logger.AspNetCore.

I don't know how else to say that I am NOT using this package. His response states:
"Is there a reason you want to use this library versus using the logging built into Lambda to CloudWatch logs?"

I AM using the logging built into Lambda to CloudWatch logs.

**Comment by ashishdhingra** [comment]
Hi @mhelbig-oz,

Thanks for your input, my apologies for overlooking that you are still using the built in logging. Please confirm if you are instantiating a new instance of LambdaLogger (just reconfirming) or using it from context.LambdaLogger.Log. The note specifies using directly from the context. Is it possible to share some code snippet (with sensitive information removed) so that I could have a look?

Thanks,
Ashish

**Comment by mhelbig-oz** [comment]
I am using the same function call as referenced in this [example](https://docs.amazonaws.cn/en_us/lambda/latest/dg/csharp-logging.html). Amazon.Lambda.Core.LambdaLogger.Log, which is a static function provided by the AWS sdk.

In my original question I mentioned that I wasn't sure if there was an actual difference between that and the logger provided on the context. If I need to use the one on the context, then I will, but I'd love to know if there is an actual difference before I make the code change.

**Comment by ashishdhingra** [comment]
@mhelbig-oz Thanks. I have posted this question for development team and will post an update here based on their suggestion.

**Comment by ashishdhingra** [comment]
Hi @mhelbig-oz,

Good morning.

As per input received from dev team, the Amazon.Lambda.Core.LambdaLogger.Log should work fine. ILambdaContext.Logger is also the same implementation behind the scenes and switching to it might not make any difference.

If this rare log miss is still to be considered an issue, please open a new issue with bug label in [aws-lambda-dotnet](https://github.com/aws/aws-lambda-dotnet) repository along with the code snippets, any stack trace etc. and dev team would prioritize to have a look at it.

Thanks,
Ashish
